libcalico-go/lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -210,6 +251,43 @@ func (c *blockAffinityClient) DeleteKVP(ctx context.Context, kvp *model.KVPair)
 	return c.toV1(kvp)
 }
 
+func (c *blockAffinityClient) deleteKVPV3(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	// We need to mark as deleted first, since the Kubernetes API doesn't support
+	// compare-and-delete. This update operation allows us to eliminate races with other clients.
+	var err error
+	nkvp := kvp
+	if kvp.Value == nil {
+		// Need to check if a value is given since V3 deletes can be made by providing a key only.
+		// Look up missing values with the provided key.
+		nkvp, err = c.getV3(ctx, kvp.Key.(model.ResourceKey), kvp.Revision)
+		if err != nil {
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+				return nil, fmt.Errorf("Unable to find block affinity. Block affinity may have already been deleted.")
+			}
+			return nil, fmt.Errorf("Error retrieving block affinity for deletion: %s", err)
+		}
+	}
+
+	nkvp.Value.(*libapiv3.BlockAffinity).Spec.Deleted = fmt.Sprintf("%t", true)
```

Comment:
I would explicitly set it:

```
nkvp.Revision = kvp.Revision
```

Reply:
There is a little bit of an unintended consequence of this as well. Since the call for `Delete` from the v3 client can be only a name, the `ResourceVersion/Revision` may not be available in this call (though it always is in calls from the v1 client. Since the `ResourceVersion/Revision` is needed for the race condition checks (and the use of this call is intended to be internal anyways), I made `ResourceVersion/Revision` required if this is called from the v3 client on a Kubernetes datastore cluster. This should be only be used internally anyways so it's not a huge deal, but it's something to make note of.

Reply:
> Since the call for Delete from the v3 client can be only a name

The v3 client allows specification of a revision. If it's not present, the semantic is "Don't perform a compare-and-delete" and if it is present, the semantic is "use the revision to perform a CAD"

Reply:
In other words, it doesn't need to be mandatory, but clients generally want to specify it.